fontbe/src/features.rs

```diff
@@ -219,13 +317,153 @@ impl<'a> FeatureWriter<'a> {
+    fn add_marks(&self, builder: &mut FeatureBuilder) -> Result<(), Error> {
+        let mut anchors = HashMap::new();
+        for (work_id, arc_glyph_anchors) in self.raw_anchors {
+            let glyph_anchors: &GlyphAnchors = arc_glyph_anchors;
+            let FeWorkId::Anchor(glyph_name) = work_id else {
+                return Err(Error::ExpectedAnchor(work_id.clone()));
+            };
+            anchors.insert(glyph_name.clone(), glyph_anchors);
+        }
+
+        let mark_groups = create_mark_groups(&anchors);
+
+        // Build the actual mark base and mark mark constructs using fea-rs builders
+
+        let mut mark_base_lookups = Vec::new();
+        let mut mark_mark_lookups = Vec::new();
+
+        for (group_name, group) in mark_groups.iter() {
+            // if we have bases *and* marks produce mark to base
+            if !group.bases.is_empty() && !group.marks.is_empty() {
+                for (base_name, base_anchor) in group.bases.iter() {
+                    let base_gid = self.glyph_id(base_name)?;
+                    let mark_class: SmolStr = format!("MC_{}", base_anchor.name).into();
+
+                    let mut mark_base = MarkToBaseBuilder::default();
+                    for (mark_name, mark_anchor) in group.marks.iter() {
+                        let mark_gid = self.glyph_id(mark_name)?;
+                        let mark_anchor_table = self.create_anchor_table(mark_anchor, builder)?;
+                        mark_base
+                            .insert_mark(mark_gid, mark_class.clone(), mark_anchor_table.clone())
+                            .map_err(Error::PreviouslyAssignedClass)?;
+                    }
+
+                    let base_anchor_table = self.create_anchor_table(base_anchor, builder)?;
+                    mark_base.insert_base(base_gid, &mark_class, base_anchor_table);
+
+                    // each in it's own lookup, whch differs from fontmake
+                    mark_base_lookups.push(builder.add_lookup(
+                        LookupFlag::default(),
+                        None,
+                        vec![mark_base],
+                    ));
+                }
+            }
+
+            // If a mark has anchors that are themselves marks what we got here is a mark to mark
+
+            let mut mark_mark = MarkToMarkBuilder::default();
+            let mut filter_set = Vec::new();
+
+            for (mark_name, mark_anchor) in group.marks.iter() {
+                let Some(glyph_anchors) = anchors.get(mark_name) else {
+                    continue;
+                };
+                if !glyph_anchors
+                    .anchors
+                    .iter()
+                    .any(|a| matches!(AnchorInfo::new(&a.name), AnchorInfo::Mark(..)))
+                {
+                    continue;
+                }
+                let mark_gid = self.glyph_id(mark_name)?;
+                let base_name = AnchorInfo::new(&mark_anchor.name).group_name();
```

Comment:
isn't this `base_name` the same as `group_name` local variable defined in the outer for loop above?

Reply:
Yes.

Reply:
Fixed.